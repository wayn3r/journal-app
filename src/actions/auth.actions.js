import Swal from 'sweetalert2'
import { AUTH_TYPES as types } from 'types'
import { authentication, signin, signup, updateProfile } from 'db/firebase.config'
import { clearStore } from 'actions'

export const startLoginEmailPassword = (email, password) => {
    return async dispatch => {
        dispatch(authLoading())
        try {
            const { user } = await signin.email(email, password)
            dispatch(login(user))
        } catch (error) {
            Swal.fire('Error', 'Incorrect user or password', 'error')
        }
        dispatch(authStop())
    }
}
export const startGoogleLogin = () => {
    return async dispatch => {
        dispatch(authLoading())
        try {
            const { user } = await signin.popup()
            dispatch(login(user))
        } catch (error) {
            throw error
        }
        dispatch(authStop())
    }
}

export const startSignUp = (name, email, password) => {
    return async dispatch => {
        dispatch(authLoading())
        try {
            const { user } = await signup.createUser(email, password)
            await updateProfile(user, { displayName: name })
            dispatch(login(user))
        } catch (error) {
            Swal.fire('Error', error.message, 'error')
        }
        dispatch(authStop())
    }
}

export const authLoading = () => ({
    type: types.loading,
})
export const authStop = () => ({
    type: types.finished,
})
export const login = ({ uid, displayName, photoURL }) => ({
    type: types.login,
    payload: {
        uid,
        name: displayName,
        photo: photoURL,
    },
})
export const logout = () => ({
    type: types.logout,
})
export const startLogOut = () => {
    return async dispatch => {
        await authentication.logout()
        dispatch(logout())
        dispatch(clearStore())
    }
}
