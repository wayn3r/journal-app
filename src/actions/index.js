import types from 'types'

export * from './auth.actions'
export * from './notes.actions'
export const clearStore = () => ({
    type: types.clearStore,
})
