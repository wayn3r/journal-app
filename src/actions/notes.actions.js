import Swal from 'sweetalert2'
import { db } from 'db/firebase.config'
import { uploadImage } from 'services/cloudinary'
import { NOTES_TYPES as types } from 'types'

export const startAddNote = () => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth
        const note = {
            title: '',
            body: '',
            date: new Date().getTime(),
        }
        const { id } = await db.addDoc(uid + '/journal/notes', note)
        dispatch(setNote(id, note))
        dispatch(addNote(id, note))
    }
}

export const setNote = (id, note) => ({
    type: types.active,
    payload: { ...note, id },
})
export const addNote = (id, note) => ({
    type: types.add,
    payload: { ...note, id },
})

export const startLoadNotes = uid => {
    return async dispatch => {
        try {
            const notes = (await db.getDocs(`${uid}/journal/notes`)).docs.map(doc => ({
                ...doc.data(),
                id: doc.id,
            }))
            dispatch(loadNotes(notes))
        } catch (err) {
            throw err
        }
    }
}
export const loadNotes = notes => ({
    type: types.load,
    payload: notes,
})

export const startSaveNote = note => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth
        const { id, ...noteToFireStore } = note
        try {
            await db.updateDoc(uid + '/journal/notes/' + id, noteToFireStore)
            dispatch(refreshNotes(id, noteToFireStore))
            Swal.fire('Note', 'Note saved successfully', 'success')
        } catch (e) {
            Swal.fire('Note', 'An error has ocurred while trying to update your note', 'error')
        }
    }
}

export const refreshNotes = (id, note) => ({
    type: types.update,
    payload: { ...note, id },
})

export const startUploadNoteImage = image => {
    return async (dispatch, getState) => {
        const { note } = getState().notes

        Swal.fire({
            title: 'Uploading image',
            text: 'Please wait...',
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading()
            },
        })
        const url = await uploadImage(image)
        const newNote = { ...note, image: url }
        dispatch(setNote(newNote.id, newNote))
        dispatch(startSaveNote(newNote))
        Swal.close()
    }
}

export const startDeleteNote = id => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth

        await db.deleteDoc(uid + '/journal/notes/' + id)

        dispatch(deleteNote(id))
    }
}
export const deleteNote = id => ({
    type: types.delete,
    payload: id,
})
