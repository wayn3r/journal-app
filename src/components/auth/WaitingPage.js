export const PleaseWaitPage = () => {
    return (
        <div className='please-wait'>
            <div className='please-wait__title'>Journal App</div>
            <div className='please-wait__message'>
                <i className='please-wait__icon'></i>
                <span className='please-wait__text'>Checking authentication, please wait...</span>
            </div>
        </div>
    )
}
