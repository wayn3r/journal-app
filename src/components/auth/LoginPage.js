import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from 'hooks'
import { Form, Input } from '../forms'
import { GoogleLoginButton } from '../buttons/GoogleLoginButton'
import { startGoogleLogin, startLoginEmailPassword } from 'actions'
import { check } from 'helpers/check'
const validations = {
    email: ({ value, email }) => [
        check(!email.isEmpty(value), 'Email is required'),
        check(email.isEmail(value), 'Email is invalid'),
    ],
    password: ({ value, password }) => [check(!password.isEmpty(value), 'Password is required')],
}

export const LoginPage = () => {
    const dispatch = useDispatch()
    const { loading } = useSelector(state => state.auth)
    const [values, onChange, , validate] = useForm(
        {
            email: '',
            password: '',
        },
        validations
    )
    const { email, password } = values

    const handleLogin = e => {
        if (!validate(e)) return

        dispatch(startLoginEmailPassword(email, password))
    }
    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin())
    }
    return (
        <div className='auth__container animate__animated animate__fadeIn'>
            <h1 className='auth__title'>Login</h1>
            <Form className='auth__form' onSubmit={handleLogin}>
                <Input
                    label='Email'
                    type='email'
                    value={email}
                    name='email'
                    invalid='Invalid email'
                    onChange={onChange}
                />
                <Input
                    type='password'
                    label='Password'
                    name='password'
                    value={password}
                    onChange={onChange}
                />
                <button className='button button--primary' disabled={loading} type='submit'>
                    {!loading ? (
                        'Login'
                    ) : (
                        <>
                            <i className='fa fa-spinner fa-spin'></i> Loading...
                        </>
                    )}
                </button>
            </Form>
            <div className='auth__social-networks'>
                <p>Login with social networks</p>
                <GoogleLoginButton onClick={handleGoogleLogin} />
            </div>

            <Link to='/auth/signup' className='link auth__link'>
                Create new account
            </Link>
        </div>
    )
}
