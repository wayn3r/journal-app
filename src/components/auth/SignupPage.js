import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { startSignUp } from 'actions'
import { useForm } from 'hooks'
import { check } from 'helpers/check'
import { Form, Input } from '../forms'

const initial = {
    name: '',
    email: '',
    password: '',
    password2: '',
}
const validations = {
    name: ({ value, name }) => [
        check(!name.isEmpty(value), 'Name is required'),
        check(name.isLength(value, { min: 3 }), 'Name is too short'),
    ],
    email: ({ value, email }) => [
        check(!email.isEmpty(value), 'Email is required'),
        check(email.isEmail(value), 'Email is invalid'),
    ],
    password: ({ value, password }) => [
        check(!password.isEmpty(value), 'Password is required'),
        check(password.isLength(value, { min: 6 }), 'Password must be at least 6 characters'),
    ],
    password2: ({ value, form }) => [check(value === form.password, 'Passwords do not match')],
}

export const SignupPage = () => {
    const [values, onChange, , validate] = useForm(initial, validations)
    const { name, email, password, password2 } = values
    const dispatch = useDispatch()
    const { loading } = useSelector(state => state.auth)
    const handleSignin = e => {
        if (!validate(e)) return

        dispatch(startSignUp(name, email, password))
    }

    return (
        <div className='auth__container animate__animated animate__fadeIn'>
            <h1 className='auth__title'>Sign Up</h1>
            <Form className='auth__form' onSubmit={handleSignin}>
                <Input
                    label='Name'
                    type='text'
                    name='name'
                    value={name}
                    invalid='Invalid name'
                    onChange={onChange}
                />
                <Input
                    label='Email'
                    type='email'
                    name='email'
                    value={email}
                    invalid='Invalid email'
                    onChange={onChange}
                />
                <Input
                    label='Password'
                    type='password'
                    name='password'
                    value={password}
                    invalid='Invalid password'
                    onChange={onChange}
                />
                <Input
                    label='Confirm password'
                    type='password'
                    name='password2'
                    value={password2}
                    invalid='Invalid password'
                    onChange={onChange}
                />

                <button className='button button--primary' type='submit' disabled={loading}>
                    {!loading ? (
                        'Sign Up'
                    ) : (
                        <>
                            <i className='fa fa-spinner fa-spin'></i> Loading...
                        </>
                    )}
                </button>
                <Link to='/auth/login' className='link auth__link'>
                    Already registered?
                </Link>
            </Form>
        </div>
    )
}
