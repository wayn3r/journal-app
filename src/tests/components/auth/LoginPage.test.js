import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { MemoryRouter } from 'react-router-dom'

import { LoginPage } from 'components/auth/LoginPage'
import { startGoogleLogin, startLoginEmailPassword } from 'actions/auth.actions'
jest.mock('actions/auth.actions', () => ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn(),
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initialState = {
    auth: {},
    notes: {},
}
const store = mockStore(initialState)
store.dispatch = jest.fn()
const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <LoginPage />
        </MemoryRouter>
    </Provider>
)

describe('Pruebas en <LoginPage />', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })
    test('debe renderizarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('debe de llamar la acción startGoogleLogin', () => {
        wrapper.find('GoogleLoginButton').prop('onClick')()
        expect(startGoogleLogin).toHaveBeenCalled()
    })
    test('debe de llamar la acción startLoginEmailPassword', () => {
        const formEvent = { target: { elements: {} } }
        wrapper.find('Form').prop('onSubmit')(formEvent)
        expect(startLoginEmailPassword).toHaveBeenCalledWith('', '')
    })
})
