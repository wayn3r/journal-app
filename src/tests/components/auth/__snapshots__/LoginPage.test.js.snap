// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginPage /> debe renderizarse correctamente 1`] = `
<div
  className="auth__container animate__animated animate__fadeIn"
>
  <h1
    className="auth__title"
  >
    Login
  </h1>
  <form
    className="auth__form"
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="input-group"
    >
      <input
        autoComplete="none"
        className="input-group__input"
        name="email"
        onChange={[Function]}
        placeholder=" "
        type="email"
        value=""
      />
      <label
        className="input-group__label"
      >
        Email
      </label>
      <span
        className="input-group__invalid-text"
      >
        Invalid email
      </span>
    </div>
    <div
      className="input-group"
    >
      <input
        autoComplete="none"
        className="input-group__input"
        name="password"
        onChange={[Function]}
        placeholder=" "
        type="password"
        value=""
      />
      <label
        className="input-group__label"
      >
        Password
      </label>
      <span
        className="input-group__invalid-text"
      >
        Password
      </span>
    </div>
    <button
      className="button button--primary"
      type="submit"
    >
      Login
    </button>
  </form>
  <div
    className="auth__social-networks"
  >
    <p>
      Login with social networks
    </p>
    <button
      className="google-button"
      onClick={[Function]}
    >
      <div
        className="google-icon-wrapper"
      >
        <img
          alt="google button"
          className="google-icon"
          src="https://www.google.com/images/branding/googleg/1x/googleg_standard_color_128dp.png"
        />
      </div>
      <p
        className="button-text"
      >
        <b>
          Sign in with google
        </b>
      </p>
    </button>
  </div>
  <a
    className="link auth__link"
    href="/auth/signup"
    onClick={[Function]}
  >
    Create new account
  </a>
</div>
`;
